WordPress/gutenberg
# Duotone Tracking & History

## #26751 Add duotone to image block<br>#26752 Add duotone block supports

⚠️ Do not merge this PR—the ones linked above should be be the ones merged. ⚠️

Since this is a rather large PR with a long history, I've simplified it down into two smaller PRs, #26751 and #26752, that can be run and reviewed independently of one another.

Both PRs share a few commits for the duotone component, toolbar, and theme.json code. I'll continue to add code to this PR which has all the history and apply those changes to the individual commits on the respective smaller PR.

#26751, the image block PR, is a bit more straightforward, so I recommend reviewing that first. #26752 requires some additional changes to core in WordPress/wordpress-develop#984, and it involves more blocks.

## Description


Adds duotone filters as a block-supports feature—originally prototyped on [Automattic/block-experiments](https://github.com/Automattic/block-experiments/tree/master/blocks/duotone-filter). With the SVG approach, there is [wide browser support](https://caniuse.com/mdn-svg_elements_fecomponenttransfer), so it'll also be supported on IE11.

The duotone feature comes with its own default palette. And a new option for `duotone` has been added to `theme.json`. See [theme-json.md](https://github.com/ajlende/gutenberg/blob/add/duotone-filter/docs/designers-developers/developers/themes/theme-json.md#settings) for details.

Adding it to a block is done with the `supports.duotone` option in the `block.json`. See [block-supports.md](https://github.com/ajlende/gutenberg/blob/add/duotone-filter/docs/designers-developers/developers/block-api/block-supports.md#duotone) for details.

In the future this can be expanded to more than two colors. That's mostly a UI problem right now, the feature supports arrays with more than two colors on the backend.

Support for the image block is added in this PR. ~Adding it to the cover block is blocked by #25171.~ Additionally the video block, cover block, and media & text block have support added via block supports which depends on WordPress/wordpress-develop#984.

## How has this been tested?




1. Create an image block, video block, cover block, and media & text block
2. Apply duotone filter from presets, the bar, and the palette selectors
3. Save and preview the duotone blocks
4. Convert between the supported blocks to see that duotone is maintained across the transformation
5. Test that you can add duotone with keyboard only

## Screenshots 

<details>
<summary>Old screenshots</summary>
<p>

<img width="586" alt="duotone toolbar and picker" src="https://user-images.githubusercontent.com/5129775/107091425-deca5980-67a5-11eb-9dc4-d885831a74f0.png">

<img width="586" alt="duotone in the cover block" src="https://user-images.githubusercontent.com/5129775/107092507-d541f100-67a7-11eb-829e-016034fe8958.png">

<img width="586" alt="duotone in the image block" src="https://user-images.githubusercontent.com/5129775/107092739-34076a80-67a8-11eb-8c05-780a38292a35.png">

<img width="586" alt="duotone in the media & text block" src="https://user-images.githubusercontent.com/5129775/107092868-5dc09180-67a8-11eb-99d3-fa2928d43556.png">

<img width="648" alt="duotone in the video block" src="https://user-images.githubusercontent.com/5129775/107420963-8d7bdc00-6abd-11eb-924e-557b4c15ed2c.png">

![media-text-duotone-2020-11-05](https://user-images.githubusercontent.com/5129775/98304629-910f3c80-1f7d-11eb-8f8a-47a399801b7e.gif)

![image-duotone-2020-11-05](https://user-images.githubusercontent.com/5129775/98301774-91f19f80-1f78-11eb-83ef-bc1fafb4d7de.gif)

![duotone](https://user-images.githubusercontent.com/5129775/96730511-76c94200-1373-11eb-90ff-452201b4d071.gif)

</p>
</details>
